program-metadata/program/src/state.rs

```diff
@@ -0,0 +1,43 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+
+/// prefix used for PDAs to avoid certain collision attacks (https://en.wikipedia.org/wiki/Collision_attack#Chosen-prefix_collision_attack)
+pub const CLASS_PREFIX: &str = "program_metadata";
+
+// Metadata size
+pub const MAX_NAME_LENGTH: usize = 36;
+
+pub const MAX_VALUE_LENGTH: usize = 256;
+
+pub const METADATA_ENTRY_SIZE: usize = 1 + MAX_NAME_LENGTH + MAX_VALUE_LENGTH;
+
+// Idl size
+pub const MAX_URL_LENGTH: usize = 200;
+
+pub const IDL_HASH_SIZE: usize = 32;
+
+pub const VERSIONED_IDL_SIZE: usize = 1 + 32 + MAX_URL_LENGTH + IDL_HASH_SIZE + MAX_URL_LENGTH;
+
+#[repr(C)]
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone)]
+pub enum AccountType {
+    MetadataPairV1,
+    VersionedIdlV1,
+}
+
+#[repr(C)]
+#[derive(Clone, BorshSerialize, BorshDeserialize, Debug)]
+pub struct MetadataEntry {
+    pub account_type: AccountType,
+    pub name: String,
+    pub value: String,
+}
+
+#[repr(C)]
+#[derive(Clone, BorshSerialize, BorshDeserialize, Debug)]
+pub struct VersionedIdl {
+    pub account_type: AccountType,
+    pub effective_slot: u64,
+    pub idl_url: String,
+    pub idl_hash: [u8; 32],
+    pub source_url: String,
```

Comment:
Is the intent that explorers could use the source to verify the deployed code? We might want to include bpf sdk version here as well as expected bytecode hash?